File: model-forge/packages/reconciler/src/smart-router.ts
import * as crypto from "crypto";
import {
  SmartRouterBackend,
  SmartRouterRouteSpec,
} from "@modelforge/benchmark-schema";

export interface RouteResult {
  route: SmartRouterRouteSpec;
  backend: SmartRouterBackend;
}

export class SmartRouter {
  private readonly backends: Map<string, SmartRouterBackend> = new Map();
  // Prefix hash to Set of backend IDs that hold the warm KV-cache
  private readonly prefixCacheMap: Map<string, Set<string>> = new Map();

  constructor(initialBackends: SmartRouterBackend[] = []) {
    for (const b of initialBackends) {
      this.registerBackend(b);
    }
  }

  public registerBackend(backend: SmartRouterBackend): void {
    this.backends.set(backend.id, { ...backend });
    for (const hash of backend.warm_prefix_hashes || []) {
      this.addPrefixAffinity(hash, backend.id);
    }
  }

  public getBackend(id: string): SmartRouterBackend | undefined {
    return this.backends.get(id);
  }

  public listBackends(): SmartRouterBackend[] {
    return Array.from(this.backends.values());
  }

  public updateHeartbeat(id: string): void {
    const b = this.backends.get(id);
    if (b) {
      b.last_heartbeat = new Date().toISOString();
    }
  }

  /**
   * Fast-path routing algorithm (<1ms) featuring:
   * 1. Prefix-cache affinity to reuse warm GPU KV-cache.
   * 2. Least-connection balancing among non-draining backends.
   */
  public route(
    prompt: string,
    model: string,
  ): RouteResult {
    const startNs = process.hrtime.bigint();
    const requestId = `req-${crypto.randomUUID().slice(0, 8)}`;

    // Generate deterministic prompt prefix hash (first 64 characters / chunk block)
    const prefixSegment = prompt.trim().slice(0, 64);
    const prefixHash = crypto
      .createHash("sha256")
      .update(prefixSegment)
      .digest("hex")
      .slice(0, 16);

    // 1. Check prefix affinity for warm KV cache
    const affinityCandidates = this.prefixCacheMap.get(prefixHash);
    let selectedBackend: SmartRouterBackend | null = null;
    let cacheHit = false;

    if (affinityCandidates && affinityCandidates.size > 0) {
      for (const candidateId of affinityCandidates) {
        const b = this.backends.get(candidateId);
        if (
          b &&
          b.status === "active" &&
          b.model === model &&
          b.active_requests < b.capacity
        ) {
          selectedBackend = b;
          cacheHit = true;
          break;
        }
      }
    }

    // 2. Fallback to least-connection active backend
    if (!selectedBackend) {
      const activeBackends = Array.from(this.backends.values()).filter(
        (b) => b.status === "active" && b.model === model,
      );

      if (activeBackends.length === 0) {
        throw new Error(
          `No active healthy inference backend available for model ${model}`,
        );
      }

      // Sort by active requests ascending
      activeBackends.sort((a, b) => a.active_requests - b.active_requests);
      selectedBackend = activeBackends[0]!;
    }

    // Update state: increment active requests and bind prefix
    selectedBackend.active_requests += 1;
    this.addPrefixAffinity(prefixHash, selectedBackend.id);

    const endNs = process.hrtime.bigint();
    const overheadMs = Number(endNs - startNs) / 1_000_000.0;

    const routeSpec: SmartRouterRouteSpec = {
      request_id: requestId,
      selected_worker_id: selectedBackend.id,
      cache_hit: cacheHit,
      prefix_hash: prefixHash,
      routing_overhead_ms: Math.round(overheadMs * 1000) / 1000,
      spot_drain_migrated: false,
    };

    return { route: routeSpec, backend: selectedBackend };
  }

  /**
   * Initiates graceful spot instance drain upon preemption notice.
   * Immediately stops new request routing and migrates active workload safely.
   */
  public drainWorker(
    workerId: string,
    reason = "spot_instance_preemption",
  ): {
    worker_id: string;
    status: string;
    reason: string;
    active_requests_remaining: number;
    migrated_count: number;
  } {
    const worker = this.backends.get(workerId);
    if (!worker) {
      throw new Error(`Worker ${workerId} not found`);
    }

    worker.status = "draining";

    // Remove worker from prefix cache routing affinity
    for (const [hash, workers] of this.prefixCacheMap.entries()) {
      workers.delete(workerId);
      if (workers.size === 0) {
        this.prefixCacheMap.delete(hash);
      }
    }

    // Migrate any active requests to alternative active workers
    let migratedCount = 0;
    if (worker.active_requests > 0) {
      const activeAlternative = Array.from(this.backends.values()).find(
        (b) => b.id !== workerId && b.status === "active" && b.model === worker.model,
      );

      if (activeAlternative) {
        activeAlternative.active_requests += worker.active_requests;
        migratedCount = worker.active_requests;
        worker.active_requests = 0;
      }
    }

    return {
      worker_id: workerId,
      status: "draining",
      reason,
      active_requests_remaining: worker.active_requests,
      migrated_count: migratedCount,
    };
  }

  public completeRequest(workerId: string): void {
    const b = this.backends.get(workerId);
    if (b && b.active_requests > 0) {
      b.active_requests -= 1;
    }
  }

  private addPrefixAffinity(prefixHash: string, workerId: string): void {
    if (!this.prefixCacheMap.has(prefixHash)) {
      this.prefixCacheMap.set(prefixHash, new Set());
    }
    this.prefixCacheMap.get(prefixHash)!.add(workerId);

    const b = this.backends.get(workerId);
    if (b && !b.warm_prefix_hashes.includes(prefixHash)) {
      b.warm_prefix_hashes.push(prefixHash);
    }
  }
}

File: model-forge/packages/benchmark-cli/modelforge/smart_router.py
"""ModelForge Ultra-Low-Latency Smart Router Client & Benchmarker.

Prefix-cache affinity dispatch, spot instance drain simulation, and sub-millisecond
routing overhead measurements.
"""

from __future__ import annotations

import hashlib
import time
import uuid
from typing import Any


class SmartRouterClient:
    """Simulates or connects to ModelForge Smart Router (<1ms overhead)."""

    def __init__(self, router_url: str = "http://localhost:3000/api/v1/router") -> None:
        self.router_url = router_url
        self.backends: list[dict[str, Any]] = [
            {
                "id": "worker-h100-east-01",
                "model": "meta-llama/Llama-3-70B",
                "status": "active",
                "active_requests": 0,
                "warm_prefixes": set(),
            },
            {
                "id": "worker-h100-east-02",
                "model": "meta-llama/Llama-3-70B",
                "status": "active",
                "active_requests": 4,
                "warm_prefixes": set(),
            },
        ]

    def route_request(self, prompt: str, model: str = "meta-llama/Llama-3-70B") -> dict[str, Any]:
        """Calculates fast-path route with prefix-cache affinity and measures overhead."""
        start_ns = time.perf_counter_ns()
        prefix_chunk = prompt.strip()[:64]
        prefix_hash = hashlib.sha256(prefix_chunk.encode("utf-8")).hexdigest()[:16]

        # 1. Prefix cache hit check
        selected_backend = None
        cache_hit = False

        for b in self.backends:
            if b["status"] == "active" and b["model"] == model and prefix_hash in b["warm_prefixes"]:
                selected_backend = b
                cache_hit = True
                break

        # 2. Least-connection fallback
        if not selected_backend:
            actives = [b for b in self.backends if b["status"] == "active" and b["model"] == model]
            if not actives:
                raise RuntimeError(f"No active backend available for model {model}")
            actives.sort(key=lambda x: x["active_requests"])
            selected_backend = actives[0]

        selected_backend["active_requests"] += 1
        selected_backend["warm_prefixes"].add(prefix_hash)

        end_ns = time.perf_counter_ns()
        overhead_ms = (end_ns - start_ns) / 1_000_000.0

        return {
            "request_id": f"req-{uuid.uuid4().hex[:8]}",
            "selected_worker_id": selected_backend["id"],
            "cache_hit": cache_hit,
            "prefix_hash": prefix_hash,
            "routing_overhead_ms": round(overhead_ms, 3),
            "spot_drain_migrated": False,
        }

    def trigger_spot_drain(self, worker_id: str, reason: str = "spot_termination_notice") -> dict[str, Any]:
        """Simulates immediate graceful spot drain with zero dropped requests."""
        target = next((b for b in self.backends if b["id"] == worker_id), None)
        if not target:
            raise ValueError(f"Worker {worker_id} not found")

        target["status"] = "draining"
        target["warm_prefixes"].clear()

        migrated = target["active_requests"]
        target["active_requests"] = 0

        # Migrate to alternative
        alt = next((b for b in self.backends if b["id"] != worker_id and b["status"] == "active"), None)
        if alt:
            alt["active_requests"] += migrated

        return {
            "worker_id": worker_id,
            "status": "draining",
            "reason": reason,
            "migrated_requests": migrated,
            "drain_completed": True,
        }
